test/core/client_channel/lb_policy/pick_first_test.cc

```diff
@@ -37,19 +39,80 @@ class PickFirstTest : public LoadBalancingPolicyTest {
  protected:
   PickFirstTest() : lb_policy_(MakeLbPolicy("pick_first")) {}
 
+  static RefCountedPtr<LoadBalancingPolicy::Config> MakePickFirstConfig(
+      bool shuffleAddressList) {
+    return MakeConfig(Json::FromArray({Json::FromObject({{
+        "pick_first",
+        Json::FromObject(
+            {{"shuffleAddressList", Json::FromBool(shuffleAddressList)}}),
+    }})}));
+  }
+
+  template <typename C>
+  absl::string_view GetConnectionRequestedAddress(
```

Comment:
I don't think this is the right approach here.  What this seems to be doing is checking to see which address pick_first starts connecting to first, and then sending a new address list before that connection attempt even finishes.  I think all this really checks is that pick_first will pick *one* of the addresses first, but it does not actually check the order in which it attempts the other addresses in the list if the connection attempt on the first address fails.  And since the tests don't care *which* address the policy chooses first each time, I think these tests would actually pass even if the shuffling functionality was not implemented.

I've sent you #33255 to add a couple of tests showing how pick_first handles multiple addresses, which I hope will be useful examples for you to work from.  That will hopefully give you some good examples to work from here.

I think what we want to test here is the following:
- Send only a single address list update at the start of the test.  It should not be necessary to update the address list after that.
- We want to do multiple connection attempt passes.  Each pass, we should record the order in which pick_first tries the addresses.  The connection attempt for every address except the last one should fail, so that we can record the full order in which it is trying the addresses.  The last address it tries should succeed.  Then we have that subchannel indicate the connection fails by reporting IDLE, which will cause pick_first to report IDLE.
- Just like in the `GoesIdleWhenConnectionFailsThenCanReconnect` test that I added in #33255, calling `ExpectStateAndQueuingPicker(GRPC_CHANNEL_IDLE)` will trigger the pick_first policy to try to reconnect, thus starting a new connection attempt pass.  As per my comment elsewhere, it should re-shuffle the address list right before it starts to connect again.
- We can go through 3 connection attempt passes.  Then we can compare the order in which pick_first tried the addresses in each pass.  If it is shuffling the order, then those lists should be in a different order in each pass; if it's not, then they should be identical in each pass.